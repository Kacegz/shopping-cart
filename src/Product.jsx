import { useState } from "react";
import styles from "./Product.module.css";
const Product = ({ item, cart, setCart }) => {
  const [amount, setAmount] = useState(0);
  function increaseAmount() {
    setAmount(amount + 1);
  }
  function decreaseAmount() {
    if (amount > 0) setAmount(amount - 1);
  }
  function changeValue(value) {
    setAmount(value);
  }
  function handleAddToCart(item, amount) {
    if (amount > 0) {
      let cartCopy = cart.filter((cartItem) => cartItem.title !== item.title);
      setCart([...cartCopy, { ...item, amount: amount }]);
    }
  }
  return (
    <>
      <div className={styles.product}>
        <img
          src={item.image}
          alt="Product image"
          srcSet=""
          className={styles.productImg}
        />
        <div className={styles.productName}>
          <p>{item.title}</p>
          <p>Price: {item.price}$</p>
        </div>
        <div className={styles.options}>
          <div className={styles.numbers}>
            <input
              className={styles.amountButton}
              type="button"
              value="-"
              onClick={() => {
                decreaseAmount();
              }}
            />
            <input
              className=""
              type="text"
              name=""
              id="amount"
              value={amount}
              onChange={(e) => {
                changeValue(e.target.value);
              }}
            />
            <input
              className={styles.amountButton}
              type="button"
              value="+"
              onClick={() => {
                increaseAmount();
              }}
            />
          </div>
          <input
            className={styles.amountButton}
            type="button"
            value="Add to cart"
            onClick={() => {
              handleAddToCart(item, amount);
            }}
          />
        </div>
      </div>
    </>
  );
};
export default Product;
