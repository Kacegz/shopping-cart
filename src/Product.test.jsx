import { screen, render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import Product from "./Product";
import { userEvent } from "@testing-library/user-event";
async function apiFetch() {
  const response = await fetch(
    "https://fakestoreapi.com/products/category/jewelery"
  );
  return await response.json();
}
const json = await apiFetch();

describe("Product buttons", () => {
  it("Renders correctly", async () => {
    const { container } = await render(<Product item={json[0]} />);
    expect(container).toMatchSnapshot();
  });
  it("Increments amount", async () => {
    const user = userEvent.setup();
    render(<Product item={json[0]} />);
    const button = screen.getByRole("button", { name: "+" });
    await user.click(button);
    expect(screen.getByRole("textbox", { name: "" })).toHaveValue("1");
  });
  it("Decrements amount", async () => {
    const user = userEvent.setup();
    render(<Product item={json[0]} />);
    const increment = screen.getByRole("button", { name: "+" });
    const decrement = screen.getByRole("button", { name: "-" });
    await user.click(increment);
    await user.click(increment);
    await user.click(decrement);
    expect(screen.getByRole("textbox", { name: "" })).toHaveValue("1");
  });
});
