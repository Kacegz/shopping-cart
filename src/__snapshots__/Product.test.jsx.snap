// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product buttons > Renders correctly 1`] = `
<div>
  <div
    class="_product_fa50cc"
  >
    <img
      alt="Product image"
      class="_productImg_fa50cc"
      src="https://fakestoreapi.com/img/71pWzhdJNwL._AC_UL640_QL65_ML3_.jpg"
      srcset=""
    />
    <div
      class="_details_fa50cc"
    >
      <div
        class="_productName_fa50cc"
      >
        <p>
          John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet
        </p>
        <p>
          Price: 
          695
          $
        </p>
      </div>
      <div
        class="_options_fa50cc"
      >
        <div
          class="_numbers_fa50cc"
        >
          <input
            class="_amountButton_fa50cc"
            type="button"
            value="-"
          />
          <input
            class="_numberInput_fa50cc"
            id="amount"
            name=""
            type="text"
            value="0"
          />
          <input
            class="_amountButton_fa50cc"
            type="button"
            value="+"
          />
        </div>
        <input
          class="_amountButton_fa50cc"
          type="button"
          value="Add to cart"
        />
      </div>
    </div>
  </div>
</div>
`;
